subprojects/core/src/main/java/org/gradle/caching/internal/controller/DefaultBuildCacheController.java

```diff
@@ -17,108 +17,236 @@
+        BuildCacheTempFileStore tempFileStore,
         boolean logStackTraces
     ) {
-        this.local = toHandle(config.local, config.localPush, BuildCacheServiceRole.LOCAL, buildOperationExecutor, logStackTraces);
+        this.buildOperationExecutor = buildOperationExecutor;
+        this.legacyLocal = toHandle(config.legacyLocal, config.legacyLocalPush, BuildCacheServiceRole.LOCAL, buildOperationExecutor, logStackTraces);
         this.remote = toHandle(config.remote, config.remotePush, BuildCacheServiceRole.REMOTE, buildOperationExecutor, logStackTraces);
-        this.temporaryFileProvider = temporaryFileProvider;
+        this.local = toHandle(config.local, config.localPush);
+        this.tmp = tempFileStore;
     }
 
     @Nullable
     @Override
-    public <T> T load(BuildCacheLoadCommand<T> command) {
-        T metadata = null;
+    public <T> T load(final BuildCacheLoadCommand<T> command) {
+        final Unpack<T> unpack = new Unpack<T>(command);
+
         if (local.canLoad()) {
-            metadata = local.doLoad(command);
+            local.load(command.getKey(), unpack);
+            if (unpack.result != null) {
+                return unpack.result.getMetadata();
+            }
         }
-        if (metadata == null && remote.canLoad()) {
-            metadata = remote.doLoad(command);
+
+        if (legacyLocal.canLoad() || remote.canLoad()) {
+            tmp.allocate(command.getKey(), new Action<File>() {
+                @Override
+                public void execute(File file) {
+                    LoadTarget loadTarget = new LoadTarget(file);
+                    if (legacyLocal.canLoad()) {
+                        legacyLocal.load(command.getKey(), loadTarget);
+                    }
+
+                    if (remote.canLoad() && !loadTarget.isLoaded()) {
+                        remote.load(command.getKey(), loadTarget);
+                    }
+
+                    if (loadTarget.isLoaded()) {
+                        unpack.execute(file);
+                        if (local.canStore()) {
+                            local.store(command.getKey(), file);
+                        }
+                    }
+                }
+            });
+        }
+
+        BuildCacheLoadCommand.Result<T> result = unpack.result;
+        if (result == null) {
+            return null;
+        } else {
+            return result.getMetadata();
         }
-        return metadata;
     }
 
-    @Override
-    public void store(BuildCacheStoreCommand command) {
-        boolean localStore = local.canStore();
-        boolean remoteStore = remote.canStore();
-
-        if (localStore && remoteStore) {
-            doStoreBoth(command);
-        } else if (localStore) {
-            local.doStore(command);
-        } else if (remoteStore) {
-            remote.doStore(command);
+    private class Unpack<T> implements Action<File> {
+        private final BuildCacheLoadCommand<T> command;
```

Comment:
From the controller client's PoV, it's a load operation. I think this name shouldn't be changed.